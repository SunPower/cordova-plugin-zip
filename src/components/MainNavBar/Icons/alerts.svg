<svg width="13" height="19" viewBox="0 0 13 19" fill="none" xmlns="http://www.w3.org/2000/svg">
<g opacity="0.5">
<path d="M10.7872 12.0611C10.483 12.0611 10.234 12.345 10.234 12.6919C10.234 13.0388 10.483 13.3227 10.7872 13.3227C11.3957 13.3227 11.8936 13.8904 11.8936 14.5843V14.7673C11.8936 15.4612 11.3957 16.0289 10.7872 16.0289H2.21277C1.60426 16.0289 1.10638 15.4612 1.10638 14.7673V14.5843C1.10638 13.8904 1.60426 13.3227 2.21277 13.3227H8.29787C8.60213 13.3227 8.85106 13.0388 8.85106 12.6919C8.85106 12.345 8.60213 12.0611 8.29787 12.0611H2.76596V8.14376C2.76596 5.77822 4.44766 3.86056 6.52213 3.86056C8.5966 3.86056 10.2783 5.77822 10.2783 8.14376V10.2507C10.2783 10.5976 10.5272 10.8815 10.8315 10.8815C11.1357 10.8815 11.3847 10.5976 11.3847 10.2507V8.14376C11.3847 5.08433 9.20511 2.59894 6.52213 2.59894C3.83915 2.59894 1.65957 5.08433 1.65957 8.14376V12.1494C0.685957 12.4333 0.00553191 13.4363 0 14.5843V14.7673C0 16.1614 0.990213 17.2905 2.21277 17.2905H4.53617C4.83489 18.5269 5.95234 19.2586 7.0366 18.918C7.73362 18.7035 8.27574 18.0853 8.46383 17.2905H10.7872C12.0098 17.2905 13 16.1614 13 14.7673V14.5843C13 13.1902 12.0098 12.0611 10.7872 12.0611ZM6.5 17.7384C6.19574 17.7384 5.91362 17.5744 5.7366 17.2905H7.2634C7.08638 17.5744 6.80426 17.7384 6.5 17.7384ZM5.94681 1.46979V0.63081C5.94681 0.283865 6.19574 0 6.5 0C6.80426 0 7.05319 0.283865 7.05319 0.63081V1.46979C7.05319 1.81673 6.80426 2.1006 6.5 2.1006C6.19574 2.1006 5.94681 1.81673 5.94681 1.46979Z" fill="white"/>
</g>
</svg>
