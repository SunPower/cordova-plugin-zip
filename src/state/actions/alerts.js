import { createAction } from 'redux-act'
import { httpGet } from '../../shared/fetch'

export const ALERTS_FETCH_INIT = createAction('ALERTS_FETCH_INIT')
export const ALERTS_FETCH_SUCCESS = createAction('ALERTS_FETCH_SUCCESS')
export const ALERTS_FETCH_ERROR = createAction('ALERTS_FETCH_ERROR')

let alertsPollTimer
const ALERTS_POLL_INTERVAL = 1000 * 60

export const pollAlerts = () => {
  return (dispatch, getState) => {
    const callback = async () => {
      const state = getState()
      const isLoggedIn = !!state.user.auth.tokenID

      if (alertsPollTimer && !isLoggedIn) {
        clearInterval(alertsPollTimer)
        return
      }

      const addressId = state.user.auth.addressId

      try {
        dispatch(ALERTS_FETCH_INIT())
        const { status, data } = await httpGet(
          `/address/${addressId}/alerts`,
          state
        )
        return status === 200
          ? dispatch(ALERTS_FETCH_SUCCESS(data))
          : dispatch(ALERTS_FETCH_ERROR(data))
      } catch (err) {
        dispatch(ALERTS_FETCH_ERROR(err))
      }
    }

    if (!alertsPollTimer) {
      callback()
      alertsPollTimer = setInterval(callback, ALERTS_POLL_INTERVAL)
    }
  }
}
export const ALERTS_STOP_POLLING_INIT = createAction('ALERTS_STOP_POLLING_INIT')
export const ALERTS_STOP_POLLING_SUCCESS = createAction(
  'ALERTS_STOP_POLLING_SUCCESS'
)

export const stopPollingAlerts = () => {
  return dispatch => {
    dispatch(ALERTS_STOP_POLLING_INIT())
    if (alertsPollTimer) {
      clearInterval(alertsPollTimer)
      dispatch(ALERTS_STOP_POLLING_SUCCESS())
      return
    }
  }
}

export const ALERTS_DISMISS = createAction('ALERTS_DISMISS')
export const ALERTS_SEEN = createAction('ALERTS_SEEN')
