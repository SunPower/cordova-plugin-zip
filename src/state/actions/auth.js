import { createAction } from 'redux-act'
import { getUser } from './user'
import { httpPost, httpDelete, httpGet } from '../../shared/fetch'
import { trimObject } from '../../shared/trim'

export const WORKFLOW_TYPES = {
  SELF: 1,
  THIRD_PARTY: 2,
  INSTALLER: 3
}

export const LOGIN_INIT = createAction('LOGIN_INIT')
export const LOGIN_SUCCESS = createAction('LOGIN_SUCCESS')
export const LOGIN_ERROR = createAction('LOGIN_ERROR')

export const performLogin = values => {
  return async dispatch => {
    try {
      dispatch(LOGIN_INIT())
      const response = await httpPost('/authenticate', values)
      if (response.status !== 200) {
        dispatch(LOGIN_ERROR(response))
        return
      }
      dispatch(LOGIN_SUCCESS(response.data))
      dispatch(getUser())
    } catch (err) {
      dispatch(LOGIN_ERROR(err))
    }
  }
}

export const CREATE_ACCOUNT_INIT = createAction('CREATE_ACCOUNT_INIT')
export const CREATE_ACCOUNT_SUCCESS = createAction('CREATE_ACCOUNT_SUCCESS')
export const CREATE_ACCOUNT_ERROR = createAction('CREATE_ACCOUNT_ERROR')

export const createAccount = values => {
  return async dispatch => {
    try {
      dispatch(CREATE_ACCOUNT_INIT())
      const response = await httpPost('/register/createAccount', {
        EmailAddress: values.email,
        FirstName: values.name.substr(0, values.name.indexOf(' ')),
        LastName: values.name.substr(values.name.indexOf(' ')),
        Password: values.password,
        Phone: values.phoneNumber,
        SerialNumber: values.serial,
        AddressID: values.addressId,
        TermsAgree: values.termsAgree
      })
      return response.status === 200
        ? dispatch(CREATE_ACCOUNT_SUCCESS(response.data))
        : dispatch(CREATE_ACCOUNT_ERROR(response))
    } catch (err) {
      dispatch(CREATE_ACCOUNT_ERROR(err))
    }
  }
}

export const LOGOUT = createAction('LOGOUT')

export const logout = () => {
  return async (dispatch, getState) => {
    const state = getState()
    await httpDelete('/session', state)
    dispatch(LOGOUT())
  }
}

export const sendResetPasswordEmail = async email => {
  return httpGet(`/passwordreset/sendPasswordResetEmail?emailaddress=${email}`)
}

export const validatePackage = async pkg => {
  return httpGet(`/passwordreset/validatePackage?package=${pkg}`)
}

export const RESET_PASSWORD_INIT = createAction('RESET_PASSWORD_INIT')
export const RESET_PASSWORD_SUCCESS = createAction('RESET_PASSWORD_SUCCESS')
export const RESET_PASSWORD_ERROR = createAction('RESET_PASSWORD_ERROR')

export const resetPassword = (values, pkg) => {
  return async dispatch => {
    try {
      dispatch(RESET_PASSWORD_INIT())
      const payload = {
        ConfirmEmailAddress: values.email,
        NewPassword: values.password,
        package: pkg,
        TermsAgree: values.termsAgree
      }
      const response = await httpPost(
        '/passwordreset/resetPassword',
        trimObject(payload)
      )
      if (response.status !== 200) {
        dispatch(RESET_PASSWORD_ERROR(response))
        return
      }
      dispatch(RESET_PASSWORD_SUCCESS())
      dispatch(
        performLogin(
          trimObject({
            username: values.email,
            password: values.password
          })
        )
      )
    } catch (err) {
      dispatch(RESET_PASSWORD_ERROR(err))
    }
  }
}

export const checkEmail = async email => {
  return httpPost(`/register/checkEmail`, {
    EmailAddress: email,
    WorkflowType: WORKFLOW_TYPES.SELF
  })
}

export const validateSerial = async serial => {
  return httpGet(`/register/validateSerial/${serial}`)
}

export const VALIDATE_SESSION_INIT = createAction('VALIDATE_SESSION_INIT')
export const VALIDATE_SESSION_SUCCESS = createAction('VALIDATE_SESSION_SUCCESS')
export const VALIDATE_SESSION_ERROR = createAction('VALIDATE_SESSION_ERROR')

export const validateSession = () => {
  return async (dispatch, getState) => {
    try {
      const state = getState()
      dispatch(VALIDATE_SESSION_INIT())
      const response = await httpGet('/session', state)
      if (response.status !== 200) {
        dispatch(VALIDATE_SESSION_ERROR(response))
        dispatch(LOGOUT())
        return
      }
      dispatch(VALIDATE_SESSION_SUCCESS())
    } catch (err) {
      dispatch(VALIDATE_SESSION_ERROR(err))
    }
  }
}
