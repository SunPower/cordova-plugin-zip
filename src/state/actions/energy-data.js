import moment from 'moment'
import { createAction } from 'redux-act'
import { httpGet } from '../../shared/fetch'
import { roundDecimals } from '../../shared/rounding'

export const GET_ENERGY_DATA_INIT = createAction('GET_ENERGY_DATA_INIT')
export const GET_ENERGY_DATA_SUCCESS = createAction('GET_ENERGY_DATA_SUCCESS')
export const GET_ENERGY_DATA_ERROR = createAction('GET_ENERGY_DATA_ERROR')

export const INTERVALS = {
  HOUR: 'hour',
  DAY: 'day',
  MONTH: 'month'
}

// Minimum amount of time that has to pass before fetching more data
export const DELTAS = {
  [INTERVALS.HOUR]: 1000 * 60 * 5 /* 5 minutes */,
  [INTERVALS.DAY]: 1000 * 60 * 30 /* 30 minutes */,
  [INTERVALS.MONTH]: 1000 * 60 * 60 * 12 /* 12 hours */
}

const timers = {}

export const ENERGY_DATA_STOP_POLLING_INIT = createAction(
  'ENERGY_DATA_STOP_POLLING_INIT'
)
export const ENERGY_DATA_STOP_POLLING_SUCCESS = createAction(
  'ENERGY_DATA_STOP_POLLING_SUCCESS'
)

export const stopPollingEnergyData = () => {
  return async dispatch => {
    dispatch(ENERGY_DATA_STOP_POLLING_INIT())
    Object.values(INTERVALS).forEach(i => {
      if (timers[i]) {
        clearInterval(timers[i])
        timers[i] = null
      }
    })
    dispatch(ENERGY_DATA_STOP_POLLING_SUCCESS())
  }
}

export const pollEnergyData = (
  interval = INTERVALS.HOUR,
  defaultstarttime = moment()
    .subtract(1, 'day')
    .valueOf()
) => {
  return async (dispatch, getState) => {
    const delta = DELTAS[interval]
    const pollFn = () => {
      const state = getState()
      const currentTimestamp = moment().valueOf()
      const lastDataTimestamp =
        state.energyData &&
        state.energyData[interval] &&
        state.energyData[interval].endtime
          ? state.energyData[interval].endtime
          : defaultstarttime

      const hasOutdatedData = currentTimestamp - lastDataTimestamp >= delta

      if (hasOutdatedData) {
        dispatch(
          getEnergyData(lastDataTimestamp - delta, currentTimestamp, interval)
        )
      }
    }

    if (!timers[interval]) {
      pollFn()
      timers[interval] = setInterval(pollFn, delta)
    }
  }
}

export const getEnergyData = (
  starttime,
  endtime,
  interval = INTERVALS.HOUR
) => {
  return async (dispatch, getState) => {
    try {
      endtime = endtime || moment().valueOf()
      starttime =
        starttime ||
        moment(endtime)
          .subtract(1, 'day')
          .valueOf()

      dispatch(GET_ENERGY_DATA_INIT({ interval }))
      const state = getState()
      const params = Object.entries({
        interval,
        starttime,
        endtime
      })
        .map(v => v.join('='))
        .join('&')

      const { status, data } = await httpGet(
        `/address/${state.user.auth.addressId}/mockenergy?${params}`,
        state
      )

      if (status !== 200) {
        dispatch(GET_ENERGY_DATA_ERROR(data))
        return
      }
      dispatch(
        GET_ENERGY_DATA_SUCCESS({
          starttime,
          endtime,
          interval,
          data
        })
      )
      dispatch(normalizeEnergyData(starttime, endtime, interval, data))
    } catch (error) {
      dispatch(GET_ENERGY_DATA_ERROR({ interval, error }))
    }
  }
}

export const NORMALIZE_ENERGY_DATA_SUCCESS = createAction(
  'NORMALIZE_ENERGY_DATA_SUCCESS'
)

const parseNumber = value => roundDecimals(parseFloat(value || 0))

export const normalizeEnergyData = (starttime, endtime, interval, data) => {
  return dispatch => {
    const normalizedData = data.reduce((acc, curr) => {
      /*
        pp = power production
        pc = power consumption
        ps = storage power
        p = production
        c = consumption
        s = storage
        soc = state of charge
        weather = storage
      */
      const [timestamp, pp, pc, ps, p, c, s, soc, weather] = curr.split(',')
      return {
        ...acc,
        [timestamp]: {
          pp: parseNumber(pp),
          pc: parseNumber(pc),
          ps: parseNumber(ps),
          p: parseNumber(p),
          c: parseNumber(c),
          s: parseNumber(s),
          soc: parseNumber(soc),
          weather
        }
      }
    }, {})

    dispatch(
      NORMALIZE_ENERGY_DATA_SUCCESS({
        starttime,
        endtime,
        interval,
        normalizedData
      })
    )
  }
}
