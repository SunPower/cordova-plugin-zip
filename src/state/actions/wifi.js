import { createAction } from 'redux-act'
import { httpGet, httpPost } from '../../shared/fetch'

export const WIFI_COLLECTOR_INIT = createAction('WIFI_COLLECTOR_INIT')
export const WIFI_COLLECTOR_SUCCESS = createAction('WIFI_COLLECTOR_SUCCESS')
export const WIFI_COLLECTOR_ERROR = createAction('WIFI_COLLECTOR_ERROR')

export const getWifiCollector = (fetchStatus = true) => {
  return async (dispatch, getState) => {
    try {
      const state = getState()
      const addressId = state.user.auth.addressId
      const hasWifiInitialData = state.wifi.hasWifiInitialData
      if (!hasWifiInitialData) {
        dispatch(WIFI_COLLECTOR_INIT())
        const { status, data } = await httpGet(
          `/wifi/collector?addressId=${addressId}`,
          state
        )
        if (status === 200) {
          dispatch(WIFI_COLLECTOR_SUCCESS(data))
          if (fetchStatus) {
            dispatch(getWifiStatus(data.ComponentSerialNumber))
          }
        } else {
          dispatch(WIFI_COLLECTOR_ERROR(data))
        }
      }
    } catch (err) {
      dispatch(WIFI_COLLECTOR_ERROR(err))
    }
  }
}

export const WIFI_STATUS_INIT = createAction('WIFI_STATUS_INIT')
export const WIFI_STATUS_SUCCESS = createAction('WIFI_STATUS_SUCCESS')
export const WIFI_STATUS_ERROR = createAction('WIFI_STATUS_ERROR')

export const getWifiStatus = (serialNumber = null) => {
  return async (dispatch, getState) => {
    try {
      dispatch(WIFI_STATUS_INIT())
      const state = getState()
      serialNumber = serialNumber || state.wifi.collector.ComponentSerialNumber
      const { status, data } = await httpGet(
        `/wifi/status?serialNumber=${serialNumber}`,
        state
      )
      return status === 200
        ? dispatch(WIFI_STATUS_SUCCESS(data))
        : dispatch(WIFI_STATUS_ERROR(data))
    } catch (err) {
      dispatch(WIFI_STATUS_ERROR(err))
    }
  }
}

export const WIFI_NETWORKS_INIT = createAction('WIFI_NETWORKS_INIT')
export const WIFI_NETWORKS_SUCCESS = createAction('WIFI_NETWORKS_SUCCESS')
export const WIFI_NETWORKS_ERROR = createAction('WIFI_NETWORKS_ERROR')

export const getWifiNetworks = (serialNumber = null, scan = false) => {
  return async (dispatch, getState) => {
    try {
      dispatch(WIFI_NETWORKS_INIT())
      const state = getState()
      serialNumber = serialNumber || state.wifi.collector.ComponentSerialNumber
      const { status, data } = await httpGet(
        `/wifi/networks?serialNumber=${serialNumber}&scan=${scan}`,
        state
      )
      return status === 200
        ? dispatch(WIFI_NETWORKS_SUCCESS(data))
        : dispatch(WIFI_NETWORKS_ERROR(data))
    } catch (err) {
      dispatch(WIFI_NETWORKS_ERROR(err))
    }
  }
}

export const WIFI_SET_INIT = createAction('WIFI_SET_INIT')
export const WIFI_SET_SUCCESS = createAction('WIFI_SET_SUCCESS')
export const WIFI_SET_ERROR = createAction('WIFI_SET_ERROR')

export const setWifiNetwork = values => {
  return async dispatch => {
    try {
      dispatch(WIFI_SET_INIT())
      const { status, data } = await httpPost('/wifi/setNetwork', {
        SSID: values.SSID,
        password: values.password,
        serialNumber: values.serialNumber
      })
      return status === 200
        ? dispatch(WIFI_SET_SUCCESS(data))
        : dispatch(WIFI_SET_ERROR(data))
    } catch (err) {
      dispatch(WIFI_SET_ERROR(err))
    }
  }
}

export const WIFI_COMMAND_STATUS_INIT = createAction('WIFI_COMMAND_STATUS_INIT')
export const WIFI_COMMAND_STATUS_SUCCESS = createAction(
  'WIFI_COMMAND_STATUS_SUCCESS'
)
export const WIFI_COMMAND_STATUS_ERROR = createAction(
  'WIFI_COMMAND_STATUS_ERROR'
)

export const getWifiCommandStatus = (tokenNumber, serialNumber = null) => {
  return async (dispatch, getState) => {
    try {
      dispatch(WIFI_COMMAND_STATUS_INIT())
      const state = getState()
      serialNumber = serialNumber || state.wifi.collector.ComponentSerialNumber
      const { status, data } = await httpGet(
        `/wifi/setNetwork/status?serialNumber=${serialNumber}&tokenNumber=${tokenNumber}`,
        state
      )
      return status === 200
        ? dispatch(WIFI_COMMAND_STATUS_SUCCESS(data))
        : dispatch(WIFI_COMMAND_STATUS_ERROR(data))
    } catch (err) {
      dispatch(WIFI_COMMAND_STATUS_ERROR(err))
    }
  }
}
