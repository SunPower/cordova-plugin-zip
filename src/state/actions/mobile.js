import { createAction } from 'redux-act'
import { validateSession } from './auth'

export const DEVICE_RESUME = createAction('DEVICE_RESUME')

export const deviceResumeListener = () => {
  return (dispatch, getState) => {
    const state = getState()

    if (!state.global.isDeviceResumeListened) {
      document.addEventListener(
        'resume',
        () => {
          dispatch(DEVICE_RESUME())
          dispatch(validateSession())
        },
        false
      )
    }
  }
}
