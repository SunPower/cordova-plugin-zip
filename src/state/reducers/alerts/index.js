import { createReducer } from 'redux-act'
import {
  ALERTS_FETCH_SUCCESS,
  ALERTS_DISMISS,
  ALERTS_SEEN
} from '../../actions/alerts'
import { LOGOUT } from '../../actions/auth'

const initialState = {
  data: {}
}

export const alertsReducer = createReducer(
  {
    [ALERTS_FETCH_SUCCESS]: (state, { unresolvedAlerts }) => ({
      ...state,
      data: {
        ...state.data,
        ...unresolvedAlerts
          .map(newAlert => {
            const id = newAlert.AlertsID
            const existingAlert = state.data[id] || {}
            const seenTimestamp =
              state.data[id] && state.data[id].seenTimestamp
                ? state.data[id].seenTimestamp
                : null
            return { [id]: { ...existingAlert, ...newAlert, seenTimestamp } }
          })
          .reduce((acc, curr) => ({ ...acc, ...curr }), {})
      }
    }),
    [ALERTS_DISMISS]: (state, payload) => ({
      ...state,
      data: {
        ...state.data,
        ...{ [payload]: { ...state.data[payload], dismissed: true } }
      }
    }),
    [ALERTS_SEEN]: state => ({
      ...state,
      data: Object.keys(state.data)
        .map(key => {
          const newAlert = {
            [key]: {
              ...state.data[key],
              seenTimestamp: Date.now().valueOf()
            }
          }
          return newAlert
        })
        .reduce((acc, curr) => ({ ...acc, ...curr }), {})
    }),
    [LOGOUT]: () => initialState
  },
  initialState
)
