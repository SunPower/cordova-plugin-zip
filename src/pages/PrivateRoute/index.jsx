import moment from 'moment'
import React, { useEffect, useLayoutEffect } from 'react'
import { useDispatch } from 'react-redux'
import { Route, Redirect } from 'react-router-dom'
import { validateSession } from '../../state/actions/auth'
// import { pollAlerts } from '../../state/actions/alerts'
import { pollEnergyData, INTERVALS } from '../../state/actions/energy-data'
import { getWifiCollector } from '../../state/actions/wifi'
import { deviceResumeListener } from '../../state/actions/mobile'
import paths from '../Router/paths'

export default function PrivateRoute({ isLoggedIn, ...props }) {
  const OriginalComponent = props.component
  const location = props.location
  const dispatch = useDispatch()

  useEffect(() => {
    if (isLoggedIn) {
      dispatch(deviceResumeListener())
      dispatch(validateSession())
      // TODO: uncomment this after SPI
      // dispatch(pollAlerts())
      dispatch(getWifiCollector())
      dispatch(pollEnergyData(INTERVALS.HOUR))
      dispatch(
        pollEnergyData(
          INTERVALS.DAY,
          moment()
            .subtract(1, 'month')
            .valueOf()
        )
      )
      dispatch(
        pollEnergyData(
          INTERVALS.MONTH,
          moment()
            .subtract(1, 'year')
            .valueOf()
        )
      )
      // Add more long running services here
    }
  })

  useLayoutEffect(() => {
    window.scrollTo(0, 0)
  })

  return (
    <Route
      {...props}
      component={innerProps =>
        isLoggedIn ? (
          <OriginalComponent {...innerProps} />
        ) : (
          <Redirect
            to={{
              pathname: paths.LOGIN,
              state: {
                from:
                  location.pathname === paths.LOGIN
                    ? paths.ROOT
                    : location.pathname
              }
            }}
          />
        )
      }
    />
  )
}
