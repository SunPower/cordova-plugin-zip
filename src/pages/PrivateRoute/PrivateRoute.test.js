import React from 'react'
import * as reactRedux from 'react-redux'
import * as alertsActions from '../../state/actions/alerts'
import * as energyDataActions from '../../state/actions/energy-data'
import * as wifiActions from '../../state/actions/wifi'
import * as authActions from '../../state/actions/auth'
import * as mobileActions from '../../state/actions/mobile'
import * as i18n from '../../shared/i18n'
import PrivateRoute from '.'

describe('Private Route', () => {
  const myComponent = () => <div>Example component</div>
  let dispatchMock
  const previousLocation = {
    pathname: '/'
  }

  beforeEach(() => {
    dispatchMock = jest.fn()
    window.scrollTo = jest.fn()

    jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => dispatchMock)
    jest
      .spyOn(i18n, 'useI18n')
      .mockImplementation(path => key => key.toUpperCase())
    jest
      .spyOn(mobileActions, 'deviceResumeListener')
      .mockImplementation(() => 'device ready listener fired')
    jest
      .spyOn(alertsActions, 'pollAlerts')
      .mockImplementation(() => 'poll alerts fired')
    jest
      .spyOn(energyDataActions, 'pollEnergyData')
      .mockImplementation(() => 'poll energy data fired')
    jest
      .spyOn(wifiActions, 'getWifiCollector')
      .mockImplementation(() => 'fetch wifi initial data fired')
    jest
      .spyOn(authActions, 'validateSession')
      .mockImplementation(() => 'validate session fired')
  })

  it('redirects if not logged in', () => {
    const component = mountWithProvider(
      <PrivateRoute location={previousLocation} />
    )({})
    expect(component).toMatchSnapshot()
    expect(window.scrollTo).toBeCalled()
  })

  it('renders the provided component in a route if we are logged in', () => {
    const component = mountWithProvider(
      <PrivateRoute
        location={previousLocation}
        isLoggedIn={true}
        component={myComponent}
      />
    )({})
    expect(component).toMatchSnapshot()
    expect(window.scrollTo).toBeCalled()
  })

  it('does not actions if we are not logged in', () => {
    mountWithProvider(<PrivateRoute location={previousLocation} />)({})
    expect(dispatchMock).not.toBeCalled()
    expect(window.scrollTo).toBeCalled()
  })

  it('fires data fetch actions if we are logged in', () => {
    mountWithProvider(
      <PrivateRoute
        location={previousLocation}
        isLoggedIn={true}
        component={myComponent}
      />
    )({})
    expect(dispatchMock).toBeCalledWith('device ready listener fired')
    // TODO: uncomment this after SPI
    // expect(dispatchMock).toBeCalledWith('poll alerts fired')
    expect(dispatchMock).toBeCalledWith('poll energy data fired')
    expect(dispatchMock).toBeCalledWith('fetch wifi initial data fired')
    expect(dispatchMock).toBeCalledWith('validate session fired')
    expect(window.scrollTo).toBeCalled()
  })
})
