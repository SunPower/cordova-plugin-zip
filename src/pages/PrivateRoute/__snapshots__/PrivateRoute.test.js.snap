// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Private Route redirects if not logged in 1`] = `
Object {
  "component": <BrowserRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "REPLACE",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/login",
            "search": "",
            "state": Object {
              "from": "/",
            },
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Provider
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <PrivateRoute
          location={
            Object {
              "pathname": "/",
            }
          }
        >
          <Route
            component={[Function]}
            location={
              Object {
                "pathname": "/",
              }
            }
          >
            <component
              history={
                Object {
                  "action": "REPLACE",
                  "block": [Function],
                  "createHref": [Function],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "length": 1,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "pathname": "/login",
                    "search": "",
                    "state": Object {
                      "from": "/",
                    },
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location={
                Object {
                  "pathname": "/",
                }
              }
              match={
                Object {
                  "isExact": false,
                  "params": Object {},
                  "path": "/",
                  "url": "/",
                }
              }
            >
              <Redirect
                to={
                  Object {
                    "pathname": "/login",
                    "state": Object {
                      "from": "/",
                    },
                  }
                }
              >
                <Lifecycle
                  onMount={[Function]}
                  onUpdate={[Function]}
                  to={
                    Object {
                      "pathname": "/login",
                      "state": Object {
                        "from": "/",
                      },
                    }
                  }
                />
              </Redirect>
            </component>
          </Route>
        </PrivateRoute>
      </Provider>
    </Router>
  </BrowserRouter>,
  "store": Object {
    "clearActions": [Function],
    "dispatch": [Function],
    "getActions": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
}
`;

exports[`Private Route renders the provided component in a route if we are logged in 1`] = `
Object {
  "component": <BrowserRouter
    keyLength={0}
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/login",
            "search": "",
            "state": Object {
              "from": "/",
            },
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Provider
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <PrivateRoute
          component={[Function]}
          isLoggedIn={true}
          location={
            Object {
              "pathname": "/",
            }
          }
        >
          <Route
            component={[Function]}
            location={
              Object {
                "pathname": "/",
              }
            }
          >
            <component
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "createHref": [Function],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "length": 1,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "pathname": "/login",
                    "search": "",
                    "state": Object {
                      "from": "/",
                    },
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location={
                Object {
                  "pathname": "/",
                }
              }
              match={
                Object {
                  "isExact": false,
                  "params": Object {},
                  "path": "/",
                  "url": "/",
                }
              }
            >
              <myComponent
                history={
                  Object {
                    "action": "POP",
                    "block": [Function],
                    "createHref": [Function],
                    "go": [Function],
                    "goBack": [Function],
                    "goForward": [Function],
                    "length": 1,
                    "listen": [Function],
                    "location": Object {
                      "hash": "",
                      "pathname": "/login",
                      "search": "",
                      "state": Object {
                        "from": "/",
                      },
                    },
                    "push": [Function],
                    "replace": [Function],
                  }
                }
                location={
                  Object {
                    "pathname": "/",
                  }
                }
                match={
                  Object {
                    "isExact": false,
                    "params": Object {},
                    "path": "/",
                    "url": "/",
                  }
                }
              >
                <div>
                  Example component
                </div>
              </myComponent>
            </component>
          </Route>
        </PrivateRoute>
      </Provider>
    </Router>
  </BrowserRouter>,
  "store": Object {
    "clearActions": [Function],
    "dispatch": [Function],
    "getActions": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
}
`;
